File: frontend/shell/src/api/tasks.ts
import { api } from './client'

/**
 * 多模态任务监控（控制台 ops 视图）：管理令牌（X-Admin-Token）守卫的跨租户只读列表 + 取消。
 * 注：任务**提交**由 API 消费方用其密钥经 `/v1/tasks`（Bearer 密钥）发起，控制台不代提交。
 */

export type TaskStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'cancelled'

export interface Task {
  id: number
  org_id: number
  org_name: string
  /** video.generation / image.generation / audio.speech … */
  type: string
  model_slug: string
  status: TaskStatus
  input: unknown
  extra: unknown
  vendor_task_id: string | null
  /** 计费量纲数量，如 {second:7} / {image:4} / {call:1} */
  usage: Record<string, number> | null
  charged_amount: string | null
  error: string | null
  webhook_url: string | null
  webhook_state: string | null
  poll_count: number
  created_at: string
  started_at: string | null
  finished_at: string | null
}

/** 跨租户任务列表（倒序，管理令牌）。 */
export async function listTasks(limit = 100): Promise<Task[]> {
  const { data } = await api.get<Task[]>('/api/task/admin/tasks', { params: { limit } })
  return data
}

/** 控制台取消任务（含上游取消）。 */
export async function cancelTask(id: number): Promise<Task> {
  const { data } = await api.post<Task>(`/api/task/admin/tasks/${id}/cancel`, {})
  return data
}

File: backend/crates/entity/src/tasks.rs
//! 多模态异步任务（M5a）状态机实体。无密钥字段，可直接序列化为 API 响应。
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "tasks")]
pub struct Model {
    #[sea_orm(primary_key)]
    pub id: i64,
    pub org_id: i32,
    pub api_key_id: i32,
    pub user_id: Option<i32>,
    pub app_id: Option<i32>,
    /// 任务类型：video.generation / image.generation / audio.speech …（列名 `type`，API 输出 `type`）
    #[sea_orm(column_name = "type")]
    #[serde(rename = "type")]
    pub task_type: String,
    pub model_id: i32,
    pub model_slug: String,
    pub channel_id: Option<i32>,
    pub group_slug: Option<String>,
    pub status: TaskStatus,
    pub input: Json,
    pub extra: Option<Json>,
    pub vendor_task_id: Option<String>,
    pub usage: Option<Json>,
    pub base_amount: Option<Decimal>,
    pub charged_amount: Option<Decimal>,
    pub error: Option<String>,
    pub webhook_url: Option<String>,
    pub webhook_state: Option<String>,
    pub request_id: Option<String>,
    pub poll_count: i32,
    pub created_at: DateTimeWithTimeZone,
    pub started_at: Option<DateTimeWithTimeZone>,
    pub finished_at: Option<DateTimeWithTimeZone>,
    pub updated_at: DateTimeWithTimeZone,
}

/// 任务状态机（强类型，映射 smallint）。序列化为变体名（queued/running/…）。
#[derive(Clone, Copy, Debug, PartialEq, Eq, EnumIter, DeriveActiveEnum, Serialize, Deserialize)]
#[sea_orm(rs_type = "i16", db_type = "SmallInteger")]
#[serde(rename_all = "lowercase")]
pub enum TaskStatus {
    #[sea_orm(num_value = 1)]
    Queued,
    #[sea_orm(num_value = 2)]
    Running,
    #[sea_orm(num_value = 3)]
    Succeeded,
    #[sea_orm(num_value = 4)]
    Failed,
    #[sea_orm(num_value = 5)]
    Cancelled,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}
